libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,476 @@
+   size_t size, int numDivisions, int lag)
+{
+   std::vector<double> peakDistances(peakIndices.size());
+   const auto peakSum = std::accumulate(
+      peakValues.begin(), peakValues.end(), 0., std::plus<double> {});
+   const auto odfSamplesPerDiv = 1. * size / numDivisions;
+   std::transform(
+      peakIndices.begin(), peakIndices.end(), peakDistances.begin(),
+      [&](int peakIndex) {
+         const auto shiftedIndex = peakIndex - lag;
+         const auto closestDiv = std::round(shiftedIndex / odfSamplesPerDiv);
+         // Normalized distance between 0 and 1:
+         const auto distance = 2 *
+                               (shiftedIndex - closestDiv * odfSamplesPerDiv) /
+                               odfSamplesPerDiv;
+         return distance < 0 ? -distance : distance;
+      });
+   // Calculate the score as the sum of the distances weighted by
+   // the odf values:
+   const auto weightedAverage =
+      std::inner_product(
+         peakDistances.begin(), peakDistances.end(), peakValues.begin(), 0.,
+         [](double a, double b) { return a + b; },
+         [&](double distance, float peakValue) {
+            return distance * peakValue;
+         }) /
+      peakSum;
+   return weightedAverage;
+}
+
+// Look at the amplitude across all scores. If our ODF is that of a musical
+// rhythm, then the contrast between good and bad quantizations should make a
+// big difference.
+double GetExperimentContrastMeasure(
+   const std::unordered_map<int, MIR::OnsetQuantization>& quantizations,
+   double minError)
+{
+   const auto maxErrorIt = std::max_element(
+      quantizations.begin(), quantizations.end(),
+      [](const std::pair<int, OnsetQuantization>& a,
+         const std::pair<int, OnsetQuantization>& b) {
+         return a.second.error < b.second.error;
+      });
+   const auto amplitude = maxErrorIt->second.error - minError;
+   return amplitude;
+}
+
+struct QuantizationExperiment
+{
+   const OnsetQuantization bestFit;
+   const double score;
+};
+
+QuantizationExperiment RunQuantizationExperiment(
+   const std::vector<float>& odf, const std::vector<int>& peakIndices,
+   const std::vector<float>& peakValues,
+   const std::vector<int>& possibleNumTatums,
+   QuantizationFitDebugOutput* debugOutput)
+{
+   const auto quantizations = [&]() {
+      std::unordered_map<int, OnsetQuantization> quantizations;
+      std::transform(
+         possibleNumTatums.begin(), possibleNumTatums.end(),
+         std::inserter(quantizations, quantizations.end()), [&](int numTatums) {
+            const auto lag = GetOnsetLag(odf, numTatums);
+            const auto distance = GetQuantizationDistance(
+               peakIndices, peakValues, odf.size(), numTatums, lag);
+            return std::make_pair(
+               numTatums, OnsetQuantization { distance, lag, numTatums });
+         });
+      return quantizations;
+   }();
+
+   const auto bestFitIt = std::min_element(
```

Comment:
Is it correct to choose the hypothesis that minimizes error as the most likely correct hypothesis -- or, is there need to do something fancier with a Bayesian formula?  Which would make some assumptions about the unconditional probabilities of the hypotheses (how often will a loop have 6/8? how often 4/4?  etc.) and those data again might need to be derived from some corpus of examples we take as representative.

Reply:
e.g. find a large enough population of examples where time signature is deduced from explicit metadata, not calculation?